components/raftstore/src/store/fsm/peer.rs

```diff
@@ -2665,6 +2674,42 @@ where
         );
     }
 
+    fn on_voter_replicated_index_request(&mut self, from: &metapb::Peer) {
+        if !self.fsm.peer.is_leader() {
+            return;
+        }
+        let mut voter_replicated_idx = self.fsm.peer.get_store().last_index();
+        for (peer_id, p) in self.fsm.peer.raft_group.raft.prs().iter() {
+            let peer = find_peer_by_id(self.region(), *peer_id).unwrap();
+            if voter_replicated_idx > p.matched && !is_learner(peer) {
+                voter_replicated_idx = p.matched;
+            }
+        }
+        let mut resp = ExtraMessage::default();
+        resp.set_type(ExtraMessageType::MsgVoterReplicatedIndexResponse);
+        resp.voter_replicated_index = voter_replicated_idx;
+        self.fsm
+            .peer
+            .send_extra_message(resp, &mut self.ctx.trans, from);
+        debug!(
+            "leader responses voter_replicated_index to witness";
+            "region_id" => self.region().get_id(),
+            "witness_id" => from.id,
+            "leader_id" => self.fsm.peer.peer.get_id(),
+            "voter_replicated_index" => voter_replicated_idx,
+        );
+    }
+
+    fn on_voter_replicated_index_response(&mut self, msg: &ExtraMessage) {
+        if self.fsm.peer.is_leader() || !self.fsm.peer.is_witness() {
+            return;
+        }
+        self.ctx.apply_router.schedule_task(
+            self.region_id(),
+            ApplyTask::CheckCompact(self.region_id(), msg.voter_replicated_index),
```

Comment:
Question: Why do we need to explicitly call CheckCompact other than leverage raft log gc? In raft log gc we cannot delete some raft log if there's lag peer. But here if the lag peer's has made some progress, then we still can gc some logs. So I don't understand here why we need to call CheckCompact.

Reply:
`CheckCompact` is only used by witness. The leader still sends the compact command to all nodes as before, non-witnesses will be normal gc, only witness will judge whether to delay gc according to voter_replicated_index.